projects/ADuCM3029_demo_cn0415/src/platform_drivers.c

```diff
@@ -0,0 +1,741 @@
+
+/**
+ * @brief Initialize the SPI communication peripheral.
+ * @param desc - The SPI descriptor.
+ * @param init_param - The structure that contains the SPI parameters.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t spi_init(struct spi_desc **desc, struct spi_init_param param)
+{
+	/* Variable for storing the return code from SPI device */
+	ADI_SPI_RESULT e_spi_result;
+	struct spi_desc *dev;
+
+	dev = (struct spi_desc *)malloc(sizeof(*dev));
+	if (!dev)
+		return FAILURE;
+
+	dev->chip_select = param.chip_select;
+	dev->id = param.id;
+	dev->max_speed_hz = param.max_speed_hz;
+	dev->mode = param.mode;
+	dev->type = param.type;
+
+	/* Open the SPI device. It opens in Master mode by default */
+	if ((e_spi_result = adi_spi_Open(dev->id, master_spi_device_mem,
+					 ADI_SPI_MEMORY_SIZE, &h_spi_device)) != ADI_SPI_SUCCESS) {
+		goto error;
+	}
+	/* Set the bit rate  */
+	if ((e_spi_result = adi_spi_SetBitrate(h_spi_device, dev->max_speed_hz)) !=
+	    ADI_SPI_SUCCESS) {
+		goto error;
+	}
+
+	/* Set the chip select. */
+	/* If no GPIO has been chosen use hardware CS */
+	if (dev->chip_select != 0xFF) {
+		if(dev->id == SPI_ARDUINO) {
+			/* SPI_ARDUINO uses CS1 */
+			if ((e_spi_result = adi_spi_SetChipSelect(h_spi_device,
+					    ADI_SPI_CS1)) != ADI_SPI_SUCCESS) {
+				goto error;
+			}
+		} else {
+			/* SPI_BLE or SPI_PMOD */
+			/* SPI_BLE and SPI_PMOD use CS0 */
+			if ((e_spi_result = adi_spi_SetChipSelect(h_spi_device,
+					    ADI_SPI_CS0)) != ADI_SPI_SUCCESS) {
+				goto error;
+			}
+		}
+	} else {
+		if ((e_spi_result = adi_spi_SetChipSelect(h_spi_device,
+				    ADI_SPI_CS_NONE)) != ADI_SPI_SUCCESS) {
+			goto error;
+		}
+	}
+	adi_spi_SetContinuousMode(h_spi_device, true);
+	adi_spi_SetIrqmode(h_spi_device, true);
+	/* set Clock polarity low */
+	adi_spi_SetClockPolarity(h_spi_device, (dev->mode & 0x02) >> 1);
+	/* set Clock phase leading */
+	adi_spi_SetClockPhase(h_spi_device, (dev->mode & 0x01) >> 0);
+
+	*desc = dev;
+
+	return CN415_SUCCESS;
+error:
+	free(dev);
+	return CN415_FAILURE;
+}
+
+/**
+ * @brief Free the resources allocated by spi_init().
+ * @param desc - The SPI descriptor.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t spi_remove(spi_desc *desc)
+{
+	ADI_SPI_RESULT e_spi_result;
+	if((e_spi_result = adi_spi_Close(h_spi_device)) != ADI_SPI_SUCCESS) {
+		return CN415_FAILURE;
+	}
+
+	free(desc);
+	return CN415_SUCCESS;
+}
+
+/**
+ * @brief Write and read data to/from SPI.
+ * @param desc - The SPI descriptor.
+ * @param data - The buffer with the transmitted/received data.
+ * @param bytes_number - Number of bytes to write/read.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t spi_write_and_read(spi_desc *desc, uint8_t *data, uint8_t bytes_number)
+{
+	ADI_SPI_TRANSCEIVER m_spi_transceive;
+	ADI_SPI_RESULT e_spi_result;
+
+	m_spi_transceive.pTransmitter = NULL;
```

Comment:
the name of the functions suggests that this is also used to write to spi. data buffer could also be placed on the transmitter and after the functions data will contain the read information.